include/core/shape.h

```diff
@@ -1772,55 +1792,52 @@ class RBFShape : public Shape<dim>
-
-  /**
-   * @brief Resets the iterable nodes to all nodes
+   * @brief Swap the vector of all nodes with a likely node vector
    * @param cell A likely one where the evaluation point is located
    */
   void
-  reset_iterable_nodes(
+  swap_iterable_nodes(
     const typename DoFHandler<dim>::active_cell_iterator cell);
 
 private:
+  std::string                          filename;
   size_t                               number_of_nodes;
   std::shared_ptr<HyperRectangle<dim>> bounding_box;
-  std::vector<size_t>                  iterable_nodes;
 
-  std::map<const typename DoFHandler<dim>::cell_iterator,
-           std::shared_ptr<std::vector<size_t>>>
+  // Elements of this vector are tuples containing: the cell barycenter, the
+  // cell diameter, and the RBF nodes located inside the active cell
+  std::vector<
+    std::tuple<Point<dim>, double, std::shared_ptr<std::vector<size_t>>>>
+    iterable_nodes;
+  // Entries of this map contain vectors of tuples containing the cell
+  // barycenter, diameter and likely nodes that are in that cell. The various
+  // elements of the vector are the tuples which contain information on the RBF
+  // nodes that affect the levelset evaluation in this cell. Note: the division
+  // of RBF nodes is made by separating nodes by which active cells they are in,
+  // and then these RBF nodes groups are referenced by the cells in a subsequent
+  // cell (by using a vector of pointers). This is required to avoid repeating
+  // the nodes IDs multiple times (to keep memory requirements low); they only
+  // appear once in a vector, then this vector is used by passing its shared
+  // pointer.
+  std::map<
+    const typename DoFHandler<dim>::cell_iterator,
+    std::shared_ptr<std::vector<
+      std::tuple<Point<dim>, double, std::shared_ptr<std::vector<size_t>>>>>>
                    likely_nodes_map;
-  size_t           max_number_of_nodes;
+  size_t           max_number_of_inside_nodes;
   DoFHandler<dim> *dof_handler;
-  Point<dim>       position_precalculated;
-  Tensor<1, 3>     orientation_precalculated;
-
-  // levels_not_precalculated is used in order to not precompute and store the
-  // likely RBF nodes for the finer levels of cells in the grid. Setting this
-  // parameter is a tradeoff between having faster distance evaluations and
-  // reducing the memory footprint: increasing levels_not_precalculated
-  // increases evaluation time.
-  int    levels_not_precalculated;
+
   double maximal_support_radius;
 
 public:
-  std::vector<size_t>           nodes_id;
-  std::vector<double>           weights;
-  std::vector<Point<dim>>       nodes_positions;
-  std::vector<Point<dim>>       rotated_nodes_positions;
-  std::vector<double>           support_radii;
-  std::vector<RBFBasisFunction> basis_functions;
+  std::vector<double>     weights;
+  std::vector<double>     nodes_positions_x;
+  std::vector<double>     nodes_positions_y;
+  std::vector<double>     nodes_positions_z;
```

Comment:
Just a question, but is it really better to have std::vectors of nodes_positions? Why not have a vector of Tensor<1,dim> or Tensor<1,3>?